### DISCUSSION TITLE
How to type in Typescript the use of immer middleware with v4 createStore

### DISCUSSION THREAD

**Discussion by jonavila** [pmndrs/zustand#1281] [Q&A] [answered]
What's the proper way of using the immer middleware in Typescript if we are using createStore to use React Context?

Example
```
import { createContext, useContext } from 'react'
import { createStore, useStore } from 'zustand'

const store = createStore(...) // vanilla store without hooks

const StoreContext = createContext()

const App = () => (
  <StoreContext.Provider value={store}>
    ...
  </StoreContext.Provider>
)
```

**Comment by dai-shi** [comment]
Is it only an issue with `immer`?
Can you create a small reproduction with https://www.typescriptlang.org/play ?

In general, it would work like something like this:
```tsx
import { createContext, useContext } from 'react'
import { createStore, useStore } from 'zustand'

const store = createStore(...) // vanilla store without hooks

const StoreContext = createContext<typeof store | null>(null)

const App = () => (
  <StoreContext.Provider value={store}>
    ...
  </StoreContext.Provider>
)
```

**Comment by jonavila** [comment, marked as answer]
@dai-shi I think I may have figure it out. My first instinct was to try to import the `WithImmer` type but noticed that wasn't available as an export. Here's an example on how I got it to work

```
import { createStore, useStore } from "zustand";
import { immer } from "zustand/middleware/immer";
import { createContext, FC, useContext } from "react";

interface MyStore {
  counter: number;
  inc: () => void;
}

const intialState = {
  counter: 5
};

const createMyStore = (state: typeof intialState = intialState) => {
  return createStore<MyStore, [["zustand/immer", never]]>(
    immer((set) => ({
      ...state,
      inc: () =>
        set((state) => {
          state.counter++;
        })
    }))
  );
};

const MyStoreContext = createContext<ReturnType<typeof createMyStore> | null>(
  null
);

export const MyStoreProvider: FC<{ counter: number }> = ({
  children,
  ...state
}) => {
  const store = createMyStore(state);

  return (
    <MyStoreContext.Provider value={store}>{children}</MyStoreContext.Provider>
  );
};

export function useMyStore(): MyStore;
export function useMyStore<T>(
  selector: (store: MyStore) => T,
  equalityFn?: (left: T, right: T) => boolean
): T;
export function useMyStore<T>(
  selector?: (store: MyStore) => T,
  equalityFn?: (left: T, right: T) => boolean
) {
  const store = useContext(MyStoreContext);

  if (!store) {
    throw new Error("MyStoreContext is not provided");
  }

  return useStore(store, selector as any, equalityFn);
}

```

The missing piece was the immer mutator reference (["zustand/immer", never]). I had missed this link
https://docs.pmnd.rs/zustand/guides/typescript#middlewares-and-their-mutators-reference

Codesandbox working example: https://codesandbox.io/s/little-worker-t306nw?file=/src/store.tsx

**Reply by dai-shi** [reply]
Does this work?
```diff
-   return createStore<MyStore, [["zustand/immer", never]]>(
+   return createStore<MyStore>()(
```

**Reply by soilSpoon** [reply]
> Does this work?
> 
> ```diff
> -   return createStore<MyStore, [["zustand/immer", never]]>(
> +   return createStore<MyStore>()(
> ```

Not working on typescript 4.6.2

**Reply by dai-shi** [reply]
@soilSpoon Can you create a small reproduction with https://www.typescriptlang.org/play ?

**Reply by Eliav2** [reply]
Thanks! That worked for me:
```ts
import { create } from "zustand";
import { immer } from "zustand/middleware/immer";

export interface BearState {
  location: GeolocationPosition | null;
  setLocation: (newLocation: GeolocationPosition) => void;
}

const useBearStore = create<BearState, [["zustand/immer", never]]>(
  immer((set) => ({
    location: null,
    setLocation: (newLocation) =>
      set((state) => {
        state.location = newLocation;
      }),
  }))
);

```

**Reply by Simon-Tang** [reply]
I think this is something that's missing from the [current documentation](https://docs.pmnd.rs/zustand/integrations/immer-middleware) which doesn't include this `[["zustand/immer", never]]` type info. Would anyone who's more familiar with this be interested in updating the docs? Or I might give this a try when I get it working (I'm using TypeScript + vanilla only Zustand).

**Reply by dai-shi** [reply]
You shouldn't need `[["zustand/immer", never]]` unless you are a middleware author.
Can anyone create a repro of the issue with TypeScript Playground?

**Reply by Simon-Tang** [reply]
[This playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQZRtNOAXzgDMoIQ4ByALwFcBnGDAOwBMB6ANzeABt+GagG4AsAChQkWIjigQaKMTIUqAIgbM2XEMHbt+aAO4Z0nBUvXiJkmAE8whAIJgweLIQC8iSXH9wGADmaABccKz0IABGSjYBERiK4cxQwKxB8QHpqGiKrDDOIeEAFACUcF4AfHDcEPo2RJKSaAAeMvDIEKzMgW54BJUo6J4D6AA8ru4s2AA0cADaC5pMLBwWIIpQ6vOsaNxKALqHVSV+2ZtKJSWMaDAV1XAlCOcJ-sFhcACMAAw-s683qwkp9qAA5EHUAESN7ZVi5fKFYpPB41W4wG4zbw1Z5wAB0BO0c0CyKJaDxHzgAGpvkQymVAXSGRIyjYWu1oJ1ur0MP18OgAOrAGAACwAolAKMofLlRvy0JM+Z5ToDLFBrujUU8XjDYSTPr9-oCEsDkjQIYooca4Qi0AUip9ypU0XdMZ5ndr8YSsfMPiksRSQtTafTGaGWSIgA) has a minimal example of Zustand (vanilla/no-React) + Immer + TypeScript, working with the `createStore<AppState, [["zustand/immer", never]]>` and with a TypeScript error when only typing with `<AppState>`.

The TypeScript error technically has enough info to work out how to resolve this (i.e. by using the typing here):

```
Argument of type 'StateCreator<AppState, [], [["zustand/immer", never]]>' is not assignable to parameter of type 'StateCreator<AppState, [], [], AppState>'.
  Type 'StateCreator<AppState, [], [["zustand/immer", never]]>' is not assignable to type '{ $$storeMutators?: [] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/immer", never]] | undefined' is not assignable to type '[] | undefined'.
        Type '[["zustand/immer", never]]' is not assignable to type '[]'.
          Source has 1 element(s) but target allows only 0.(2345)
```

**Reply by dai-shi** [reply]
Here's the canonical solution: https://tsplay.dev/wja7kW

**Reply by Simon-Tang** [reply]
Thanks @dai-shi, that's super helpful and it solves the error for my codebase as well.

**Reply by robxbob** [reply]
Been stuck on this for too long and not sure if its a typescript mis typing or WithImmer type @dai-shi 
The issue lies with how a generic type is being used as the key for ages in the example below:

[tsplay](https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240730#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wFgAoUSWOAbzgGUZokBBMYAGjjSJRiRMWcAL5xsuAgC8ArgGcYKAHYATAPQA3ZcAA2OlPgDcFKtHj1QIJFFHiceAESyFy9SGAqVOpAHcURNUtrB2NyCgokAA9qeGAlAShMdCQ4djAmfiQAHgAVOCiBVTk4BSg4gHMAPjoKODglFCs5AC44HIBtAF1auBRypBbEJDRoFVzOUoq4SooRcLDyKJieCCUFVLB05iIAYVWBSPgAXiH0GAA6XlQBPfiCrJ6hInZgLLSMgSzlAE9K6oAfOAyVRITBxJAqCiVAAUwJUoPBKgAlKEItEzHA4gkkmgUu9tkgAAo4DQeazEiBgOS5fKHJBFEowMpKKo9Ar0lTFfGZXLVWg9NAAC10KiISgA-K1kGdztKMAA5CDw0JzRbomgjNbwfEsCmk+E2E409kMyYsmH88h1IUisWcHrnR0oTYfJCzVo6oh6slQClU3lIuBHPkC1brdoKFgARk6QdOGHO8kEigE0OhgeDPD4AieSGhQSgaYz1WdW0ySIroWtYfgEYJACZYyc5Rck66i0HqldMrm3i6U9kcn98yArIX053ev3y5XQ1q4HWWABmJvx1tyZOZDuZ7s5gkjsd9sufIcwieZ0uuiso+Z1IgwGRQJRwaE9OpZGZWupwI+524HC5vQNOAtB0GQkCOWhIyIKMRE-b9v1oG0dFFelVQQrI1E9JB-wKc4gOseCEO-X8CVww58JJH0QJQMCIKghs4LfRDkNQpR0JIrD+xYcjAKog0iOIn9sN4yiIH1awaLoyDoKQJcmK-BCkOFFCxQ498uK2Hj9jwgioEEn81CIm9yBEQwgA)

**Reply by dai-shi** [reply]
Can you try this?
```diff
-   const [store2] = React.useState(() => createStore<AppState<T>>(immer(() => appState)));
+   const [store2] = React.useState(() => createStore<AppState<T>>()(immer(() => appState)));
```

**Reply by robxbob** [reply]
Did it work for you? Because it didn't for me. 😔
Updated the [tsplay](https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240730#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wFgAoUSWOAbzgGUZokBBMYAGjjSJRiRMWcAL5xsuAgC8ArgGcYKAHYATAPQA3ZcAA2OlPgDcFKtHj1QIJFFHiceAESyFy9SGAqVOpAHcURNUtrB2NyCgokAA9qeGAlAShMdCQ4djAmfiQAHgAVOCiBVTk4BSg4gHMAPjoKODglFCs5AC44HIBtAF1auBRypBbEJDRoFVzOUoq4SooRcLDyKJieCCUFVLB05iIAYVWBSPgAXiH0GAA6XlQBPfiCrJ6hInZgLLSMgSzlAE9K6oAfOAyVRITBxJAqCiVAAUwJUoPBKgAlKEItEzHA4gkkmgUu9tkgAAo4DQeazEiBgOS5fKHJBFEowMpKKo9Ar0lTFfGZXLVWg9NAAC10KiISgA-K1kGdztKMAA5CDw0JzRbomgjNbwfEsCmk+E2E409kMyYsmH88h1IUisWcHrnR0oTYfJCzVo6oh6slQClU3lIuBHPkC1brdoKFgARk6QdOGHO8kEigE0OhgeDPD4AieSGhQSgaYz1WdW0ySIroWtYfgEYJACZYyc5Rck66i0HqldMrm3i6U9kcn90-mQFZC+nO71++XK6GtXA6ywAMxN+OtuTJzIdzPdnME0fjvtlz5DmGTzOl10VlHzOpEGAyKBKODQnp1LIzK11ODH3O3A4Lm9A04C0HQZCQI5aEjIgoxEL8fx-WgbR0UV6VVRCsjUT0kAAgpzmA6wEMQn8-wJPDDgIkkfVAlBwMg6CG3g98kJQtClAw0jsP7FgKKA6iDWIkjfxwviqIgfVrFo+ioJgpBl2Y79EOQ4VULFTiP24rZeP2fDCKgITfzUYjb3IERDCAA) to have what you mentioned above.

Also, the objective is to make ages' keys dependent on the list of names. (in case you know another way to do this thats compatible with immer)

**Reply by dai-shi** [reply]
I'm not sure how the context will be used in the end, but to make the tsplay happy, how about this? https://tsplay.dev/NrBZ3N

**Reply by robxbob** [reply]
Aha, thanks. I was hoping to not use that dummyStore. How would I remove it in "production", in the case that this was in a component library, wouldnt the end-developers experience the type errors?

**Reply by dai-shi** [reply]
Yeah, this seems suboptimal, and we only hope the final production build will strip it.

On the other hand, if this is a component library, I wouldn't export the context itself which can't be generics but a custom hook which can be. How does that sound?

**Reply by robxbob** [reply]
I see what you are saying. I am in fact just exporting the generic hook as well. So should i just leave the dummy context along with not exporting any contexts?

**Reply by dai-shi** [reply]
If you don't export context, you can make the context type `any`.

**Reply by robxbob** [reply]
Aha, you're absolutely right. A bit unfortunate having the any for this one component though. I have like 5 other components that have similar structures except it doesn't have complex typings like the tsplay example, so it works properly. I'm debating if I should not export context for all, make it `any`, and define the expected type in the custom hook, OR just have that dummystore for this one case.

**Reply by dai-shi** [reply]
Hi @robxbob,

Great news! I found a real solution after reading the typescript guide again: https://github.com/pmndrs/zustand/blob/66f3a029fbc4640b76c26959e01a5caa857c04dc/docs/guides/typescript.md?plain=1#L340-L358

Here you go: https://tsplay.dev/mpkM7N

```tsx
import React from "react";
import { createStore } from "zustand/vanilla";
import { immer } from "zustand/middleware/immer";

export interface AppState<T extends string> {
  names: T[];
  ages: Record<T, string>;
}

type AnyStore = ReturnType<
  typeof createStore<AppState<any>, [["zustand/immer", never]]>
>;

export const AppStoreContext = React.createContext<AnyStore | undefined>(
  undefined,
);

export interface AppStoreProviderProps<T extends string> extends AppState<T> {
  children?: React.ReactNode;
}
export const AppStoreProvider = <T extends string>({
  children,
  ...appState
}: AppStoreProviderProps<T>) => {
  const [store1] = React.useState(() => createStore(immer(() => appState)));
  const [store2] = React.useState(() =>
    createStore<AppState<T>>()(immer(() => appState)),
  );
  const [store3] = React.useState(() =>
    createStore(immer<AppState<T>>(() => appState)),
  );

  return (
    <>
      <AppStoreContext.Provider value={store1}>
        {children}
      </AppStoreContext.Provider>
      <AppStoreContext.Provider value={store2}>
        {children}
      </AppStoreContext.Provider>
      <AppStoreContext.Provider value={store3}>
        {children}
      </AppStoreContext.Provider>
    </>
  );
};
```

**Reply by robxbob** [reply]
> You shouldn't need `[["zustand/immer", never]]` unless you are a middleware author. Can anyone create a repro of the issue with TypeScript Playground?

Oh yea! I didn't need to create a dummy store! Nice solution and i get to keep my consistency by adding this one type. I guess for this situation i can ignore the comment from above?

**Reply by dai-shi** [reply]
Yeah, it's not "middleware" but you are a library author, so you may need it.

**Comment by justjish** [comment]
Incase you are nesting multiple middlewares....

```ts
import { create } from 'zustand';
import { devtools } from 'zustand/middleware';
import { immer } from 'zustand/middleware/immer';

type State = {
  count: number;
  setCount: (newCount: number) => void;
};
// Note that the order you apply the middleware correlates to the order in the type
create<State, [['zustand/devtools', never], ['zustand/immer', never]]>(
  devtools(
    immer((set) => ({
      count: 0,
      setCount: (newCount: number) => set((state) => void (state.count = newCount)),
    })),
  ),
);

```

**Reply by gandhis1** [reply]
Am I off base, but this seems like rather clunky syntax for this? So every single mutator we have, we need to surround it in a `void`? That arguably ruins a lot of the concise expressiveness that `immer` was supposed to solve in the first place.

**Reply by acao** [reply]
for anyone going down this road with `zustand/perstist`, it requires `unknown` in place of `never`

**Reply by muzhaqi16** [reply]
zustand/perstist should be state instead of never

`['zustand/persist', State]`
